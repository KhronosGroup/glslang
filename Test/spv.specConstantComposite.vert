#version 450

layout(constant_id = 200) const int my_int = 3;
layout(constant_id = 201) const float my_float = 3.14;
layout(constant_id = 202) const
    double my_double = 3.1415926535897932384626433832795;
layout(constant_id = 203) const bool my_bool = true;

// Flat struct
struct flat_struct {
  int i;
  float f;
  double d;
  bool b;
};

// Nesting struct
struct nesting_struct {
  flat_struct nested;
  vec4 v;
  int i;
};

// Expect OpSpecConstantComposite
const flat_struct my_flat_full_spec = {my_int, my_float, my_double,
                                                 my_bool};
const flat_struct my_flat_partial_spec = {30, 30.14, my_double,
                                                    my_bool};

const vec4 my_vec4_full_spec = vec4(my_float, my_float, my_float, my_float);
const vec4 my_vec4_partial_spec = vec4(my_float, my_float, 300.14, 300.14);

const nesting_struct my_nesting_full_spec = {my_flat_full_spec, my_vec4_full_spec,
                                      my_int};
const nesting_struct my_nesting_partial_spec = {my_flat_partial_spec,
                                         my_vec4_partial_spec, 3000};

const float my_array[3] = {my_float, my_float, 1.0};

layout(constant_id = 16) const int arraySize = 5;
in vec4 ucol[arraySize];

out vec4 color;
out int size;

void refer_primary_spec_const() {
  if (my_bool) color *= my_int;
}

void refer_composite_spec_const() {
  color += my_vec4_full_spec;
  color -= my_vec4_partial_spec;
}

void refer_copmosite_dot_dereference() {
  color *= my_nesting_full_spec.i;
}

void refer_composite_bracket_dereference() {
  color -= my_array[1];
}

void declare_spec_const_in_func() {
  const nesting_struct spec_const_declared_in_func = {my_flat_partial_spec,
                                               my_vec4_partial_spec, 10};
  color /= spec_const_declared_in_func.i;
}

void main() {
  refer_primary_spec_const();
  refer_composite_spec_const();
  refer_copmosite_dot_dereference();
  refer_composite_bracket_dereference();
  declare_spec_const_in_func();
}
