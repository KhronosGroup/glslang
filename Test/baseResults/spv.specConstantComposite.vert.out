spv.specConstantComposite.vert
Warning, version 450 is not yet complete; most version-specific features are present, but some are missing.


Linked vertex stage:


// Module Version 10000
// Generated by (magic number): 80001
// Id's are bound by 81

                              Capability Shader
                              Capability Float64
               1:             ExtInstImport  "GLSL.std.450"
                              MemoryModel Logical GLSL450
                              EntryPoint Vertex 4  "main" 23 78 80
                              Source GLSL 450
                              Name 4  "main"
                              Name 6  "refer_primary_spec_const("
                              Name 8  "refer_composite_spec_const("
                              Name 10  "refer_copmosite_dot_dereference("
                              Name 12  "refer_composite_bracket_dereference("
                              Name 14  "declare_spec_const_in_func("
                              Name 23  "color"
                              Name 38  "flat_struct"
                              MemberName 38(flat_struct) 0  "i"
                              MemberName 38(flat_struct) 1  "f"
                              MemberName 38(flat_struct) 2  "d"
                              MemberName 38(flat_struct) 3  "b"
                              Name 39  "nesting_struct"
                              MemberName 39(nesting_struct) 0  "nested"
                              MemberName 39(nesting_struct) 1  "v"
                              MemberName 39(nesting_struct) 2  "i"
                              Name 78  "ucol"
                              Name 80  "size"
                              Decorate 17 SpecId 203
                              Decorate 25 SpecId 200
                              Decorate 29 SpecId 201
                              Decorate 40 SpecId 202
                              Decorate 75 SpecId 16
               2:             TypeVoid
               3:             TypeFunction 2
              16:             TypeBool
              17:    16(bool) SpecConstantTrue
              20:             TypeFloat 32
              21:             TypeVector 20(float) 4
              22:             TypePointer Output 21(fvec4)
       23(color):     22(ptr) Variable Output
              24:             TypeInt 32 1
              25:     24(int) SpecConstant 3
              29:   20(float) SpecConstant 1078523331
              30:   21(fvec4) SpecConstantComposite 29 29 29 29
              33:   20(float) Constant 1133908460
              34:   21(fvec4) SpecConstantComposite 29 29 33 33
              37:             TypeFloat 64
 38(flat_struct):             TypeStruct 24(int) 20(float) 37(float) 16(bool)
39(nesting_struct):             TypeStruct 38(flat_struct) 21(fvec4) 24(int)
              40:   37(float) SpecConstant 1413754136 1074340347
              41:38(flat_struct) SpecConstantComposite 25 29 40 17
              42:39(nesting_struct) SpecConstantComposite 41 30 25
              43:     24(int) Constant 2
              48:             TypeInt 32 0
              49:     48(int) Constant 3
              50:             TypeArray 20(float) 49
              51:   20(float) Constant 1065353216
              52:          50 SpecConstantComposite 29 29 51
              53:     24(int) Constant 1
              58:     24(int) Constant 30
              59:   20(float) Constant 1106321080
              60:38(flat_struct) SpecConstantComposite 58 59 40 17
              61:     24(int) Constant 10
              62:39(nesting_struct) SpecConstantComposite 60 34 61
              73:     24(int) Constant 3000
              74:39(nesting_struct) SpecConstantComposite 60 34 73
              75:     24(int) SpecConstant 5
              76:             TypeArray 21(fvec4) 75
              77:             TypePointer Input 76
        78(ucol):     77(ptr) Variable Input
              79:             TypePointer Output 24(int)
        80(size):     79(ptr) Variable Output
         4(main):           2 Function None 3
               5:             Label
              68:           2 FunctionCall 6(refer_primary_spec_const()
              69:           2 FunctionCall 8(refer_composite_spec_const()
              70:           2 FunctionCall 10(refer_copmosite_dot_dereference()
              71:           2 FunctionCall 12(refer_composite_bracket_dereference()
              72:           2 FunctionCall 14(declare_spec_const_in_func()
                              Return
                              FunctionEnd
6(refer_primary_spec_const():           2 Function None 3
               7:             Label
                              SelectionMerge 19 None
                              BranchConditional 17 18 19
              18:               Label
              26:   20(float)   ConvertSToF 25
              27:   21(fvec4)   Load 23(color)
              28:   21(fvec4)   VectorTimesScalar 27 26
                                Store 23(color) 28
                                Branch 19
              19:             Label
                              Return
                              FunctionEnd
8(refer_composite_spec_const():           2 Function None 3
               9:             Label
              31:   21(fvec4) Load 23(color)
              32:   21(fvec4) FAdd 31 30
                              Store 23(color) 32
              35:   21(fvec4) Load 23(color)
              36:   21(fvec4) FSub 35 34
                              Store 23(color) 36
                              Return
                              FunctionEnd
10(refer_copmosite_dot_dereference():           2 Function None 3
              11:             Label
              44:     24(int) CompositeExtract 42 2
              45:   20(float) ConvertSToF 44
              46:   21(fvec4) Load 23(color)
              47:   21(fvec4) VectorTimesScalar 46 45
                              Store 23(color) 47
                              Return
                              FunctionEnd
12(refer_composite_bracket_dereference():           2 Function None 3
              13:             Label
              54:   20(float) CompositeExtract 52 1
              55:   21(fvec4) Load 23(color)
              56:   21(fvec4) CompositeConstruct 54 54 54 54
              57:   21(fvec4) FSub 55 56
                              Store 23(color) 57
                              Return
                              FunctionEnd
14(declare_spec_const_in_func():           2 Function None 3
              15:             Label
              63:     24(int) CompositeExtract 62 2
              64:   20(float) ConvertSToF 63
              65:   21(fvec4) Load 23(color)
              66:   21(fvec4) CompositeConstruct 64 64 64 64
              67:   21(fvec4) FDiv 65 66
                              Store 23(color) 67
                              Return
                              FunctionEnd
